### ISSUE TITLE
Cannot choose VS2017 version

### ISSUE THREAD

**Issue by dfederm** [Microsoft/vsts-agent#970]
## **Agent version and platform**
Version of your agent? 2.115.0
OS of the machine running the agent? Windows

## **VSTS type and version**
VisualStudio.com or On-Prem TFS? VSO
If VisualStudio.com, what is your account name? http://dfederm.visualstudio.com

## **What's not working?** 
I have Visual Studio 2017 and Visual Studio 2017 Preview installed side by side. I'm unable to specify which o the two I want to use for the agent. It seems to always choose the preview one, which I'm having other issues with (unrelated to the agent).

## **Agent and Worker's diag log**
```
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: & C:\Agent\bin\powershell\Add-MSBuildCapabilities.ps1
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Checking: hive 'LocalMachine', view 'Registry32', key name 'Software\Microsoft\MSBuild\ToolsVersions\2.0', value name 'MSBuildToolsPath'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Found String value: 'C:\Windows\Microsoft.NET\Framework\v2.0.50727\'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: ##vso[agent.capability name=MSBuild_2.0]C:\Windows\Microsoft.NET\Framework\v2.0.50727\
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] Adding 'MSBuild_2.0': 'C:\Windows\Microsoft.NET\Framework\v2.0.50727\'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Checking: hive 'LocalMachine', view 'Registry32', key name 'Software\Microsoft\MSBuild\ToolsVersions\3.5', value name 'MSBuildToolsPath'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Found String value: 'C:\Windows\Microsoft.NET\Framework\v3.5\'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: ##vso[agent.capability name=MSBuild_3.5]C:\Windows\Microsoft.NET\Framework\v3.5\
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] Adding 'MSBuild_3.5': 'C:\Windows\Microsoft.NET\Framework\v3.5\'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Checking: hive 'LocalMachine', view 'Registry32', key name 'Software\Microsoft\MSBuild\ToolsVersions\4.0', value name 'MSBuildToolsPath'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Found String value: 'C:\Windows\Microsoft.NET\Framework\v4.0.30319\'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: ##vso[agent.capability name=MSBuild_4.0]C:\Windows\Microsoft.NET\Framework\v4.0.30319\
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] Adding 'MSBuild_4.0': 'C:\Windows\Microsoft.NET\Framework\v4.0.30319\'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Checking: hive 'LocalMachine', view 'Registry32', key name 'Software\Microsoft\MSBuild\ToolsVersions\12.0', value name 'MSBuildToolsPath'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Key not found.
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Checking: hive 'LocalMachine', view 'Registry32', key name 'Software\Microsoft\MSBuild\ToolsVersions\14.0', value name 'MSBuildToolsPath'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Found String value: 'C:\Program Files (x86)\MSBuild\14.0\bin\'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: ##vso[agent.capability name=MSBuild_14.0]C:\Program Files (x86)\MSBuild\14.0\bin\
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] Adding 'MSBuild_14.0': 'C:\Program Files (x86)\MSBuild\14.0\bin\'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Testing container: 'REGISTRY::HKEY_CLASSES_ROOT\CLSID\{177F0C4A-1CD3-4DE7-A32C-71DBBB9FA36D}'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Exists.
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Adding Visual Studio setup helpers.
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Getting Visual Studio setup instances.
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Found 2 instances.
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: 
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: 
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Description   : Microsoft DevOps solution for productivity and coordination across teams of any size
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: DisplayName   : Visual Studio Enterprise 2017
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Id            : 1e6cfc2d
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: InstallDate   : System.Runtime.InteropServices.ComTypes.FILETIME
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Name          : VisualStudio/15.2.0+26430.4
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Path          : C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Version       : 15.0.26430.4
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: VersionString : 15.0.26430.4
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: 
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Description   : Microsoft DevOps solution for productivity and coordination across teams of any size
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: DisplayName   : Visual Studio Enterprise 2017
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Id            : e845f1f3
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: InstallDate   : System.Runtime.InteropServices.ComTypes.FILETIME
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Name          : VisualStudioPreview/15.3.0-Preview+26510.0
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Path          : C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Version       : 15.0.26510.0
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: VersionString : 15.0.26510.0
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: 
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: 
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: 
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: 
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Testing leaf: 'C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\15.0\Bin\MSBuild.exe'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Exists.
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: ##vso[agent.capability name=MSBuild_15.0]C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\15.0\Bin\
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] Adding 'MSBuild_15.0': 'C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\15.0\Bin\'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: ##vso[agent.capability name=MSBuild]C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\15.0\Bin\
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] Adding 'MSBuild': 'C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\15.0\Bin\'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Checking: hive 'LocalMachine', view 'Registry64', key name 'Software\Microsoft\MSBuild\ToolsVersions\2.0', value name 'MSBuildToolsPath'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Found String value: 'C:\Windows\Microsoft.NET\Framework64\v2.0.50727\'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: ##vso[agent.capability name=MSBuild_2.0_x64]C:\Windows\Microsoft.NET\Framework64\v2.0.50727\
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] Adding 'MSBuild_2.0_x64': 'C:\Windows\Microsoft.NET\Framework64\v2.0.50727\'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Checking: hive 'LocalMachine', view 'Registry64', key name 'Software\Microsoft\MSBuild\ToolsVersions\3.5', value name 'MSBuildToolsPath'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Found String value: 'C:\Windows\Microsoft.NET\Framework64\v3.5\'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: ##vso[agent.capability name=MSBuild_3.5_x64]C:\Windows\Microsoft.NET\Framework64\v3.5\
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] Adding 'MSBuild_3.5_x64': 'C:\Windows\Microsoft.NET\Framework64\v3.5\'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Checking: hive 'LocalMachine', view 'Registry64', key name 'Software\Microsoft\MSBuild\ToolsVersions\4.0', value name 'MSBuildToolsPath'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Found String value: 'C:\Windows\Microsoft.NET\Framework64\v4.0.30319\'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: ##vso[agent.capability name=MSBuild_4.0_x64]C:\Windows\Microsoft.NET\Framework64\v4.0.30319\
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] Adding 'MSBuild_4.0_x64': 'C:\Windows\Microsoft.NET\Framework64\v4.0.30319\'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Checking: hive 'LocalMachine', view 'Registry64', key name 'Software\Microsoft\MSBuild\ToolsVersions\12.0', value name 'MSBuildToolsPath'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Key not found.
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Checking: hive 'LocalMachine', view 'Registry64', key name 'Software\Microsoft\MSBuild\ToolsVersions\14.0', value name 'MSBuildToolsPath'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Found String value: 'C:\Program Files (x86)\MSBuild\14.0\bin\amd64\'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: ##vso[agent.capability name=MSBuild_14.0_x64]C:\Program Files (x86)\MSBuild\14.0\bin\amd64\
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] Adding 'MSBuild_14.0_x64': 'C:\Program Files (x86)\MSBuild\14.0\bin\amd64\'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Testing leaf: 'C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\15.0\Bin\amd64\MSBuild.exe'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: Exists.
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: ##vso[agent.capability name=MSBuild_15.0_x64]C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\15.0\Bin\amd64\
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] Adding 'MSBuild_15.0_x64': 'C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\15.0\Bin\amd64\'
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] STDOUT: ##vso[agent.capability name=MSBuild_x64]C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\15.0\Bin\amd64\
[2017-05-14 18:39:39Z INFO PowerShellCapabilitiesProvider] Adding 'MSBuild_x64': 'C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\15.0\Bin\amd64\'
```

**Comment by ericsciple** [comment]
@dfederm the agent just chooses latest.

The agent uses the location for capability registration only.

What is the actual problem? That the capability is registered in an undesired fashion? Or that the task uses the undesired version?

A few potential mitigations/solutions depending on your actual problem:
1) Remove offending version from the machine
2) Add an additional "user" capability for routing.
3) Use task inputs to choose the correct version - e.g. use MSBuild task instead of VSBuild task and then you can specify the location to MSBuild.

**Comment by ericsciple** [comment]
Let me know if that doesn't solve and we can reopen.

**Comment by dfederm** [comment]
@ericsciple the problem is that there's no way for me to choose which version I'd like to use. I do see in the build task I can choose the VS version, but within a VS version I have no way to choose 2017 vs the Preview. Since I'm using my dev machine as an agent (I don't imagine this is uncommon for small projects), I'd like to have the Preview installed to play around with it, but I don't want it to be a candidate for the VSTS agent, or if it is I'd like to explicitly choose it.

Perhaps the VS Preview should be detected as a separate version so the VS Build task can distinguish between the two and the existing version option can list it separately.

I ended up taking the first option you provided as a mitigation, but that's less than ideal as now I can't actually use VS2017 Preview. I'd like to help test it if I can.

I'm not sure what you mean by the 2nd option, but it does sound like another thing which could work, which would be allowing users to manually set/override capabilities on the agent. I previously went to the agent pool page in VSTS to see if I could just set any of the value and also looked at the agent command line to see if there was an option to set any agent capabilities from there, but couldn't find anything. Allowing users to manually override auto-detected values seems like it could also help this situation.

The 3rd option seems to be a bit hacky if I'm understanding it correctly. It seems like you're suggesting I set the full path to MSBuild in the MSBuild task, which would be something agent-specific like `C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin\amd64\`. I feel like that breaks the entire concept of capabilities which is in part to abstract the full locations of things away form the tasks and let them be agent-agnostic (different agents can have stuff in different places).
